File: frontend/src/api/ads.ts
import api from './index'

export interface AdsNetworkStatusHourlyMetric {
  stat_date: string
  hour: number
  total_roads: number
  smooth_roads: number
  basically_smooth_roads: number
  light_congested_roads: number
  moderate_congested_roads: number
  severe_congested_roads: number
  smooth_pct: number
  basically_smooth_pct: number
  light_congested_pct: number
  moderate_congested_pct: number
  severe_congested_pct: number
  network_avg_speed: number | null
}

export interface AdsCongestionHourlyMetric {
  stat_date: string
  hour: number
  light_congested_roads: number
  moderate_congested_roads: number
  severe_congested_roads: number
  avg_congestion: number | null
  total_delay_min: number | null
}

export interface AdsTopCongestedRoadMetric {
  rank_id: number
  stat_date: string
  hour: number
  road_id: number
  road_name: string | null
  congestion_idx: number | null
  avg_speed: number | null
  trip_count: number | null
  duration_loss: number | null
}

export interface AdsTripDistanceHourlyMetric {
  stat_date: string
  hour: number
  short_trips: number
  medium_trips: number
  long_trips: number
  avg_distance: number | null
  total_distance: number | null
  sample_count: number | null
}

export interface AdsTripSpeedHourlyMetric {
  stat_date: string
  hour: number
  avg_speed: number | null
  speed_p50: number | null
  speed_p85: number | null
  speed_p95: number | null
  overspeed_ratio: number | null
  sample_count: number | null
}

export interface AdsTripTimeslotDailyMetric {
  stat_date: string
  morning_rush: number
  daytime: number
  evening_rush: number
  night: number
  weekday_trips: number | null
  holiday_trips: number | null
  sample_count: number | null
}

export interface AdsHotspotMonitorDailyMetric {
  stat_date: string
  zone_type: string
  hotspot_count: number
  total_trip_count: number
  total_pickup_count: number
  total_dropoff_count: number
  max_hotspot_zone_id: string | null
  max_hotspot_zone_name: string | null
  max_hotspot_trip_count: number | null
}

interface ItemsResponse<T> {
  items: T[]
}

export function getAdsNetworkStatusHourly(date: string) {
  return api.get<any, ItemsResponse<AdsNetworkStatusHourlyMetric>>('/data/ads/network-status/hourly', {
    params: { date },
  })
}

export function getAdsCongestionHourly(date: string) {
  return api.get<any, ItemsResponse<AdsCongestionHourlyMetric>>('/data/ads/congestion/hourly', {
    params: { date },
  })
}

export function getAdsTopCongestedRoads(date: string, hour: number, limit = 10) {
  return api.get<any, ItemsResponse<AdsTopCongestedRoadMetric>>('/data/ads/top-congested-roads', {
    params: { date, hour, limit },
  })
}

export function getAdsTripDistanceHourly(date: string) {
  return api.get<any, ItemsResponse<AdsTripDistanceHourlyMetric>>('/data/ads/trips/distance/hourly', {
    params: { date },
  })
}

export function getAdsTripSpeedHourly(date: string) {
  return api.get<any, ItemsResponse<AdsTripSpeedHourlyMetric>>('/data/ads/trips/speed/hourly', {
    params: { date },
  })
}

export function getAdsTripTimeslotDaily() {
  return api.get<any, ItemsResponse<AdsTripTimeslotDailyMetric>>('/data/ads/trips/timeslot/daily')
}

export type HotspotMonitorTimeMode = 'all' | 'day' | 'hour'

export function getAdsHotspotMonitorDaily(timeMode: HotspotMonitorTimeMode, date?: string, hour?: number) {
  const params: Record<string, string | number> = { time_mode: timeMode }
  if (date) params.date = date
  if (hour != null) params.hour = hour
  return api.get<any, ItemsResponse<AdsHotspotMonitorDailyMetric>>('/data/ads/hotspots/monitor/daily', {
    params,
  })
}

File: backend/src/schemas/data/ads.py
from datetime import date
from typing import Any, Optional

from pydantic import BaseModel

from schemas.data.common import PageResponse


class AdsNetworkStatusHourlyMetric(BaseModel):
    stat_date: date
    hour: int
    total_roads: int
    smooth_roads: int
    basically_smooth_roads: int
    light_congested_roads: int
    moderate_congested_roads: int
    severe_congested_roads: int
    smooth_pct: float
    basically_smooth_pct: float
    light_congested_pct: float
    moderate_congested_pct: float
    severe_congested_pct: float
    network_avg_speed: Optional[float] = None


class AdsCongestionHourlyMetric(BaseModel):
    stat_date: date
    hour: int
    light_congested_roads: int
    moderate_congested_roads: int
    severe_congested_roads: int
    avg_congestion: Optional[float] = None
    total_delay_min: Optional[float] = None


class AdsTopCongestedRoadMetric(BaseModel):
    rank_id: int
    stat_date: date
    hour: int
    road_id: int
    road_name: Optional[str] = None
    congestion_idx: Optional[float] = None
    avg_speed: Optional[float] = None
    trip_count: Optional[int] = None
    duration_loss: Optional[float] = None


class AdsTripDistanceHourlyMetric(BaseModel):
    stat_date: date
    hour: int
    short_trips: int
    medium_trips: int
    long_trips: int
    avg_distance: Optional[float] = None
    total_distance: Optional[float] = None
    sample_count: Optional[int] = None


class AdsTripDistanceDailyMetric(BaseModel):
    stat_date: date
    short_trips: int
    medium_trips: int
    long_trips: int
    avg_distance: Optional[float] = None
    total_distance: Optional[float] = None
    sample_count: Optional[int] = None


class AdsTripSpeedHourlyMetric(BaseModel):
    stat_date: date
    hour: int
    avg_speed: Optional[float] = None
    speed_p50: Optional[float] = None
    speed_p85: Optional[float] = None
    speed_p95: Optional[float] = None
    overspeed_ratio: Optional[float] = None
    sample_count: Optional[int] = None


class AdsTripSpeedDailyMetric(BaseModel):
    stat_date: date
    avg_speed: Optional[float] = None
    speed_p50: Optional[float] = None
    speed_p85: Optional[float] = None
    speed_p95: Optional[float] = None
    sample_count: Optional[int] = None


class AdsTripTimeslotDailyMetric(BaseModel):
    stat_date: date
    morning_rush: int
    daytime: int
    evening_rush: int
    night: int
    weekday_trips: Optional[int] = None
    holiday_trips: Optional[int] = None
    sample_count: Optional[int] = None


class AdsHotspotMonitorDailyMetric(BaseModel):
    stat_date: date
    zone_type: str
    hotspot_count: int
    total_trip_count: int
    total_pickup_count: int
    total_dropoff_count: int
    max_hotspot_zone_id: Optional[str] = None
    max_hotspot_zone_name: Optional[str] = None
    max_hotspot_trip_count: Optional[int] = None


class AdsRoadStatusHourlyMetric(BaseModel):
    road_id: int
    road_name: Optional[str] = None
    road_class: Optional[str] = None
    stat_date: date
    hour: int
    current_speed: Optional[float] = None
    current_flow: Optional[int] = None
    congestion_idx: Optional[float] = None
    status: Optional[str] = None


class AdsRoadStatusHourlyListResponse(PageResponse):
    items: list[AdsRoadStatusHourlyMetric]


class AdsHotspotDistrictHourlyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    hour: int
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotDistrictDailyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotGridHourlyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    hour: int
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotGridHourlyListResponse(PageResponse):
    items: list[AdsHotspotGridHourlyMetric]


class AdsHotspotGridDailyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotGridDailyListResponse(PageResponse):
    items: list[AdsHotspotGridDailyMetric]


class AdsHotspotClusterHourlyMetric(BaseModel):
    cluster_id: int
    stat_date: date
    hour: int
    center_lon: Optional[float] = None
    center_lat: Optional[float] = None
    cluster_type: Optional[str] = None
    trip_count: int
    pickup_count: int
    dropoff_count: int
    avg_duration: Optional[float] = None


class AdsHotspotClusterHourlyListResponse(PageResponse):
    items: list[AdsHotspotClusterHourlyMetric]


class AdsHotspotClusterDailyMetric(BaseModel):
    cluster_id: int
    stat_date: date
    center_lon: Optional[float] = None
    center_lat: Optional[float] = None
    cluster_type: Optional[str] = None
    trip_count: int
    pickup_count: int
    dropoff_count: int
    avg_duration: Optional[float] = None


class AdsHotspotClusterDailyListResponse(PageResponse):
    items: list[AdsHotspotClusterDailyMetric]


class AdsHotspotPoiHourlyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    hour: int
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotPoiHourlyListResponse(PageResponse):
    items: list[AdsHotspotPoiHourlyMetric]


class AdsHotspotPoiDailyMetric(BaseModel):
    zone_id: str
    zone_name: Optional[str] = None
    stat_date: date
    trip_count: int
    pickup_count: int
    dropoff_count: int
    vehicle_count: int
    avg_trip_distance: Optional[float] = None
    avg_duration: Optional[float] = None


class AdsHotspotPoiDailyListResponse(PageResponse):
    items: list[AdsHotspotPoiDailyMetric]


class AdsNetworkStatusHourlyResponse(BaseModel):
    items: list[AdsNetworkStatusHourlyMetric]


class AdsCongestionHourlyResponse(BaseModel):
    items: list[AdsCongestionHourlyMetric]


class AdsTopCongestedRoadsResponse(BaseModel):
    items: list[AdsTopCongestedRoadMetric]


class AdsTripDistanceHourlyResponse(BaseModel):
    items: list[AdsTripDistanceHourlyMetric]


class AdsTripDistanceDailyResponse(BaseModel):
    items: list[AdsTripDistanceDailyMetric]


class AdsTripSpeedHourlyResponse(BaseModel):
    items: list[AdsTripSpeedHourlyMetric]


class AdsTripSpeedDailyResponse(BaseModel):
    items: list[AdsTripSpeedDailyMetric]


class AdsTripTimeslotDailyResponse(BaseModel):
    items: list[AdsTripTimeslotDailyMetric]


class AdsHotspotMonitorDailyResponse(BaseModel):
    items: list[AdsHotspotMonitorDailyMetric]


class AdsHotspotDistrictHourlyResponse(BaseModel):
    items: list[AdsHotspotDistrictHourlyMetric]


class AdsHotspotDistrictDailyResponse(BaseModel):
    items: list[AdsHotspotDistrictDailyMetric]